criu/sk-tcp.c

```diff
@@ -263,7 +263,7 @@ int dump_one_tcp(int fd, struct inet_sk_desc *sk, SkOptsEntry *soe)
 	if (sk->dst_port == 0)
 		return 0;
 
-	if (opts.tcp_close) {
+	if (opts.tcp_close && sk->state != TCP_CLOSE) {
```

Comment:
It is unclear without a comment what is going on here.

And this doesn't look right. We don't need to dump closed sockets if tcp_close is set. If you want to separated them from just created sockets, I think it would be better to introduce a new state and use it for all undumped tcp sockets.

Reply:
I don't think it's necessary to introduce a new state, the problem here is the inconsistency between dumping and restoring

- When dumping with tcp_close, we don't dump established states and TCP_CLOSE
- In restoring with tcp_close, we skip restoring established states but still try to restore TCP_CLOSE

To solve this problem, I think we need to determine whether we want to checkpoint/restore TCP_CLOSE socket when tcp_close is used. We need to dump TCP_CLOSE sockets like the current solution or we must skip restoring TCP_CLOSE sockets in restoring path.

Reply:
@minhbq-99 Thank you for working on this.

> As opts.tcp_close is about closing "established" sockets and it should not affect closed ones

I agree with @Snorch's comment above ^

> It is unclear without a comment what is going on here.

@avagin Maybe we could add a comment above the if statement?

Reply:
> I agree with @Snorch's comment above ^

Do you mean this one:

> As opts.tcp_close is about closing "established" sockets and it should not affect closed ones, I like the solution from @rst0git.

The problem with this comment is that it doesn't explain why it should work this way. We are talking about sockets for which dst_port isn't zero. It means they are all "established" but some of them have been closed...

Reply:
One reason we need to checkpoint and restore sockets that are in TCP_CLOSE state is because there might be data in the buffered socket that we need to restore.

Here is an example of a test https://github.com/rst0git/criu/commit/e45dd76a67c0f229e1d9c457a54f211be3a0ea77 that would: 

- Fail without a patch:
```
==================== Run zdtm/static/socket-tcp-close3 in h ====================
 DEP       socket-tcp-close3.d
 CC        socket-tcp-close3.o
 LINK      socket-tcp-close3
Start test
./socket-tcp-close3 --pidfile=socket-tcp-close3.pid --outfile=socket-tcp-close3.out
Run criu dump
Run criu restore
=[log]=> dump/zdtm/static/socket-tcp-close3/74/1/restore.log
------------------------ grep Error ------------------------
b'(00.000696)     74: inet: \tRestore: family AF_INET    type SOCK_STREAM    proto IPPROTO_TCP      port 44244 state TCP_CLOSE        src_addr 127.0.0.1'
b'(00.000703)     74: tcp: Restoring TCP connection'
b'(00.000706)     74: tcp: Restoring TCP connection id 8 ino 137f68'
b'(00.000711)     74: No tcp-stream-137f68.img image'
b'(00.000713)     74: Error (criu/protobuf.c:72): Unexpected EOF on (empty-image)'
b'(00.000720)     74: Error (criu/files.c:1200): Unable to open fd=5 id=0x8'
b'(00.000735) Error (criu/cr-restore.c:2470): Restoring FAILED.'
------------------------ ERROR OVER ------------------------
########### Test zdtm/static/socket-tcp-close3 FAIL at CRIU restore ############
```
- Fail if we don't restore the `TCP_CLOSE` sockets (i.e., using the solution proposed in https://github.com/checkpoint-restore/criu/commit/4720a914cd92b7dc121e270cde1b0459f9256a37)
```
==================== Run zdtm/static/socket-tcp-close3 in h ====================
 DEP       socket-tcp-close3.d
 CC        socket-tcp-close3.o
 LINK      socket-tcp-close3
Start test
./socket-tcp-close3 --pidfile=socket-tcp-close3.pid --outfile=socket-tcp-close3.out
Run criu dump
Run criu restore
Send the 15 signal to  74
Wait for zdtm/static/socket-tcp-close3(74) to die for 0.100000
########### Test zdtm/static/socket-tcp-close3 FAIL at result check ############
Test output: ================================
02:26:19.205:    74: FAIL: socket-tcp-close3.c:58: Did't receieve all buffered data (errno = 11 (Resource temporarily unavailable)
```
- And pass with the current patch:
```
=================== Run zdtm/static/socket-tcp-close3 in ns ====================
Construct root for zdtm/static/socket-tcp-close3
Start test
./socket-tcp-close3 --pidfile=socket-tcp-close3.pid --outfile=socket-tcp-close3.out
Run criu dump
Run criu restore
Send the 15 signal to  88
Wait for zdtm/static/socket-tcp-close3(88) to die for 0.100000
Removing dump/zdtm/static/socket-tcp-close3/62
=================== Test zdtm/static/socket-tcp-close3 PASS ====================
==================== Run zdtm/static/socket-tcp-close3 in h ====================
Start test
./socket-tcp-close3 --pidfile=socket-tcp-close3.pid --outfile=socket-tcp-close3.out
Run criu dump
Run criu restore
Send the 15 signal to  160
Wait for zdtm/static/socket-tcp-close3(160) to die for 0.100000
Removing dump/zdtm/static/socket-tcp-close3/160
=================== Test zdtm/static/socket-tcp-close3 PASS ====================
=================== Run zdtm/static/socket-tcp-close3 in uns ===================
Start test
./socket-tcp-close3 --pidfile=socket-tcp-close3.pid --outfile=socket-tcp-close3.out
Run criu dump
Run criu restore
Send the 15 signal to  257
Wait for zdtm/static/socket-tcp-close3(257) to die for 0.100000
Removing dump/zdtm/static/socket-tcp-close3/229
=================== Test zdtm/static/socket-tcp-close3 PASS ====================
```

Reply:
I think you don't understand basic things about --tcp-close. If an user sets this options, they agree to drop all established sockets including closed ones.

Reply:
what do you think about https://github.com/avagin/criu/commit/4181b237a1b98feddaaed441f58a836289951214?

Reply:
> If an user sets this option, they agree to drop all established sockets including closed ones.

I don't agree.

The commit message introducing this option (2c37042) explicitly states:
```
this option doesn't affect sockets with original states of TCP_LISTEN and TCP_CLOSE.
```

and the man page for `--tcp-close` doesn't say anything about closed sockets:
```
--tcp-close
 Don't dump the state of, or block, established tcp connections.
 This is useful when tcp connections are not going to be restored.
```

Why do you think closed sockets should not be restored?

Reply:
> what do you think about avagin@4181b23?

The test in https://github.com/rst0git/criu/commit/e60ee8152a8e018db763a70348a8385049942858 fails with
> 54: FAIL: socket-tcp-close3.c:58: Did't receieve all buffered data (errno = 11 (Resource temporarily unavailable))

If we want this to be the expected behaviour we should update the man page and https://criu.org/CLI/opt/--tcp-close

Reply:
@rst0git Do you remember why we decided to add --tcp-close? We have users that want to dump processes and then restore them later on another host with another IP address. And they don't want to see iptables rules that we use to block TCP connections. If we apply the fix that you suggested here, we will not be able to restore processes on a host with another IP address and we will have iptables rules after criu dumped.

```
# python test/zdtm.py run -t zdtm/static/socket-tcp-close3 -f h --sbs
userns is supported
=== Run 1/1 ================ zdtm/static/socket-tcp-close3
timens isn't supported on 5.10.28-00069-g96f0f4a40fdd-dirty
==================== Run zdtm/static/socket-tcp-close3 in h ====================
Start test
./socket-tcp-close3 --pidfile=socket-tcp-close3.pid --outfile=socket-tcp-close3.out
Pause before dump. Press Enter to continue.
Run criu dump
Pause before restore. Press Enter to continue.^Z
[1]+  Stopped                 python test/zdtm.py run -t zdtm/static/socket-tcp-close3 -f h --sbs
# ps axf | grep zdtm.py
   2315 pts/0    T      0:00              \_ python test/zdtm.py run -t zdtm/static/socket-tcp-close3 -f h --sbs
   2333 pts/0    T      0:00              |   \_ ./zdtm_ct zdtm.py
   2334 pts/0    S      0:00              |       \_ python zdtm.py
   2336 pts/0    T      0:00              |           \_ python zdtm.py
   2396 pts/0    S+     0:00              \_ grep --color=auto zdtm.py     
# nsenter -n -t 2336 iptables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         
DROP       tcp  --  localhost            localhost            mark match ! 0xc114 tcp spt:60646 dpt:cddbp-alt
DROP       tcp  --  localhost            localhost            mark match ! 0xc114 tcp spt:cddbp-alt dpt:60646

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DROP       tcp  --  localhost            localhost            mark match ! 0xc114 tcp spt:cddbp-alt dpt:60646
DROP       tcp  --  localhost            localhost            mark match ! 0xc114 tcp spt:60646 dpt:cddbp-alt
```

> If we want this to be the expected behaviour we should update the man page and https://criu.org/CLI/opt/--tcp-close

Yes, this is expected behavior and we need to update our documentation. We need to tell that --tcp-close affects all established TCP sockets.

Reply:
I see, thank you for the explanation!
@minhbq-99 would you be able to apply Andrei's fix (https://github.com/avagin/criu/commit/4181b237a1b98feddaaed441f58a836289951214) and update [Documentation/criu.txt](https://github.com/checkpoint-restore/criu/blob/criu-dev/Documentation/criu.txt)?

Reply:
Hi, I updated the pull request.

```c
if (opts.tcp_close && (ii->ie->state != TCP_CLOSE || ii->ie->dst_port != 0))
```

As I can see, `ii->ie->dst_port != 0` is guaranteed to be true when reaching `restore_one_tcp` so I simplified it to

```c
if (opts.tcp_close)
```